src/detectors/BTOF/BarrelTOFNeighborFinder.cc

```diff
@@ -0,0 +1,201 @@
+  for (const auto& dir : searchDirs) {
+    auto testCell = hitCell;
+    try {
+      _decoder->set(testCell, "x", xID + dir.first);
+      _decoder->set(testCell, "y", yID + dir.second);
+    } catch (const std::runtime_error& err) {
+      // catch overflow error
+      // ignore if invalid position ID
+      continue;
+    }
+
+    try {
+      auto pos = _converter->position(testCell);
+      if (testCell != _converter->cellID(pos))
+        continue;
+    } catch (const std::invalid_argument& err) {
+      // Ignore CellID that is invalid
+      continue;
+    }
+
+    // only look for cells that have not been searched
+    if (dp.find(testCell) == dp.end()) {
+      auto testSensorID = _getSensorID(testCell);
+      if (testSensorID == sensorID) {
+        // inside the same sensor
+        this->_findAllNeighborsInSensor(testCell, ans, dp);
+      }
+    }
+  }
+}
+
+void BarrelTOFNeighborFinder::_initWithCell(const dd4hep::rec::CellID& hitCell) {
+  if (!_decoder) {
+    // cell dimensions, cellID converter and bit decoder will be fetched on first use
+    if (_log)
+      _log->info("Retrieving cellID decoder.");
+    // retrieve segmentation class
+    const auto det = _converter->findContext(hitCell)->element;
+    auto readout   = _converter->findReadout(det);
+    auto seg       = readout.segmentation();
+    auto type      = seg.type();
+    if (type != "CartesianGridXY")
+      throw std::runtime_error("Unsupported segmentation type: " + type +
+                               ". BarrelTOF must use CartesianGridXY.");
+    // retrieve meaning of cellID bits
+    _decoder = seg.decoder();
+  }
+}
+
+const std::shared_ptr<std::vector<dd4hep::rec::CellID>>
+BarrelTOFNeighborFinder::findAllNeighborInSensor(const dd4hep::rec::CellID& hitCell) {
+  _initWithCell(hitCell);
+  // look for cache. If exist, return cache value
+  if (_useCache) {
+    auto it = _cache.find(hitCell);
+    if (it != _cache.end())
+      return it->second;
+  }
+
+  std::unordered_set<dd4hep::rec::CellID> dp;
+  // fill cache
+  std::shared_ptr<std::vector<dd4hep::rec::CellID>> neighbors =
+      std::make_shared<std::vector<dd4hep::rec::CellID>>();
+  this->_findAllNeighborsInSensor(hitCell, neighbors, dp);
+
+  if (_useCache)
+    for (auto cell : *neighbors)
+      _cache[cell] = neighbors;
+
+  return neighbors;
+}
+
+const dd4hep::rec::CellID
+BarrelTOFNeighborFinder::_getSensorID(const dd4hep::rec::CellID& hitCell) {
+  _initWithCell(hitCell);
+  // fix x-y, what you left with are ids that corresponds to sensor info
+  // cellID may change when position changes.
+  auto sensorID = hitCell; //_converter -> cellID(_converter -> position(hitCell));
+  _decoder->set(sensorID, "x", 0);
+  _decoder->set(sensorID, "y", 0);
+
+  return sensorID;
+}
+
+/****************************************
+ * Functions below aren't needed anymore to find neighbors
+ * They are still handy so I keep them
+ * *************************************/
```

Comment:
Please can you remove the unused functions so it's clearer what is actually needed in the code.

Reply:
BarrelTOFNeighborFinder is now integrated into src/detector/BTOFChargeSharing.cc. All un-necessary functions are removed.